File: src/strategy/impl/FundingHarvest.ts
import type { Strategy, Signal, StrategyContext } from '../Strategy';
import type { MarketSnapshot } from '../../market/MarketDataService';

// Delta-neutral funding rate harvesting:
// High positive funding → short perp + long spot → earn funding every 8h
// High negative funding → long perp + short spot → earn funding every 8h
const FUNDING_THRESHOLD_ANNUALIZED = 0.50; // 50% APR

export const FundingHarvest: Strategy = {
  name: 'funding_harvest',
  suitableRegimes: ['trending', 'ranging', 'high_volatility'],
  evaluate(snap: MarketSnapshot, _ctx: StrategyContext): Signal {
    const fundingRate = snap.fundingRate;
    // Bybit charges funding every 8h → 3x per day → annualize: rate * 3 * 365
    const annualized = fundingRate * 3 * 365;

    if (Math.abs(annualized) < FUNDING_THRESHOLD_ANNUALIZED) {
      return { action: 'hold', symbol: snap.symbol, strategy: 'funding_harvest', confidence: 0, rationale: `Funding ${(annualized * 100).toFixed(1)}% APR below threshold` };
    }

    // Positive funding: longs pay shorts → be short perp
    if (annualized > 0) {
      return {
        action: 'enter_short',
        symbol: snap.symbol,
        strategy: 'funding_harvest',
        confidence: Math.min(0.95, 0.5 + Math.abs(annualized) / 2),
        rationale: `High positive funding ${(annualized * 100).toFixed(1)}% APR — harvest via short perp`,
        suggestedEntry: snap.markPrice,
        suggestedStop: snap.markPrice * 1.05,
      };
    }
    // Negative funding: shorts pay longs → be long perp
    return {
      action: 'enter_long',
      symbol: snap.symbol,
      strategy: 'funding_harvest',
      confidence: Math.min(0.95, 0.5 + Math.abs(annualized) / 2),
      rationale: `High negative funding ${(annualized * 100).toFixed(1)}% APR — harvest via long perp`,
      suggestedEntry: snap.markPrice,
      suggestedStop: snap.markPrice * 0.95,
    };
  },
};

File: bybit_agent/strategy/impl/funding_harvest.py
"""FundingHarvest — ports src/strategy/impl/FundingHarvest.ts."""

from __future__ import annotations

from ..base import Signal, StrategyContext

FUNDING_THRESHOLD_ANNUALIZED = 0.10  # 10% APR (~0.027% per 8h funding payment)


class FundingHarvest:
    name = "funding_harvest"
    suitable_regimes = ["trending", "ranging", "high_volatility", "crisis", "crowded_long", "crowded_short"]

    def evaluate(self, snap, ctx: StrategyContext) -> Signal:
        funding_rate = snap.fundingRate
        annualized = funding_rate * 3 * 365  # funding every 8h → 3x/day

        if abs(annualized) < FUNDING_THRESHOLD_ANNUALIZED:
            return Signal(action="hold", symbol=snap.symbol, strategy="funding_harvest",
                          confidence=0,
                          rationale=f"Funding {annualized * 100:.1f}% APR below threshold")

        if annualized > 0:
            return Signal(
                action="enter_short", symbol=snap.symbol, strategy="funding_harvest",
                confidence=min(0.95, 0.5 + abs(annualized) / 2),
                rationale=f"High positive funding {annualized * 100:.1f}% APR — harvest via short perp",
                suggestedEntry=snap.markPrice, suggestedStop=snap.markPrice * 1.05,
            )
        return Signal(
            action="enter_long", symbol=snap.symbol, strategy="funding_harvest",
            confidence=min(0.95, 0.5 + abs(annualized) / 2),
            rationale=f"High negative funding {annualized * 100:.1f}% APR — harvest via long perp",
            suggestedEntry=snap.markPrice, suggestedStop=snap.markPrice * 0.95,
        )
